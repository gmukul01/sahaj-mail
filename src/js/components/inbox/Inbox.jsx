import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import TopBar from './TopBar';
import EmailList from 'components/email/EmailList';
import ActionBar from './ActionBar';

const Inbox = props => {
	const { selectedEmails, addToSelectedEmails, clearSelectedEmails } = selectedEmailsState([]);
	const { totalEmails, emails, fetchEmails, deleteEmails, readEmails, fetchInboxDetails } = props;

	useEffect(() => {
		fetchEmails();
		fetchInboxDetails();
	}, []);

	function onRead() {
		readEmails(selectedEmails);
		clearSelectedEmails();
	}

	function onDelete() {
		deleteEmails(selectedEmails);
		clearSelectedEmails();
	}

	function onRefresh() {
		fetchEmails();
		clearSelectedEmails();
	}

	return (
		<section className="content inbox">
			<TopBar totalEmails={totalEmails} />
			<ActionBar {...{ onRefresh, onDelete, onRead }} />
			<EmailList {...{ selectedEmails, addToSelectedEmails, onRead: readEmails, emails }} />
		</section>
	);
};

Inbox.propTypes = {
	totalEmails: PropTypes.number,
	emails: PropTypes.array,
	fetchEmails: PropTypes.func,
	deleteEmails: PropTypes.func,
	readEmails: PropTypes.func,
	fetchInboxDetails: PropTypes.func
};

export default React.memo(Inbox);

const selectedEmailsState = initialValue => {
	const [emails, setValue] = useState(initialValue);

	function handleChange(e) {
		const emailId = Number(e.target.value),
			newSelectedEmails = emails.indexOf(emailId) === -1 ? [...emails, emailId] : emails.filter(id => id !== emailId);

		setValue(newSelectedEmails);
	}

	function clearSelectedEmails() {
		setValue([]);
	}

	return { selectedEmails: emails, addToSelectedEmails: handleChange, clearSelectedEmails };
};
