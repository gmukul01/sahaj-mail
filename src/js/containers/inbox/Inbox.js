import { connect } from 'react-redux';

import Inbox from 'components/inbox/Inbox';
import { fetchInboxDetails } from 'actions/inbox';
import { fetchEmails, deleteEmails, readEmails } from 'actions/emails';

const mapStateToProps = ({ inbox: { pageNumber, totalEmails, totalUnread, emailsPerPage, emails } }) => ({
	emails,
	totalEmails,
	totalUnread,
	pageNumber,
	emailsPerPage
});

const mergeProps = (stateProps, dispatchProps, ownProps) => {
	const { fetchEmails, deleteEmails, readEmails } = dispatchProps;
	const { emails, pageNumber, emailsPerPage, totalEmails, totalUnread } = stateProps;
	return {
		...stateProps,
		...dispatchProps,
		...ownProps,
		fetchEmails() {
			fetchEmails('inbox', pageNumber, emailsPerPage);
		},
		deleteEmails(selectedEmails) {
			const readEmails = emails.filter(({ id, isRead }) => selectedEmails.includes(id + '') && !isRead);
			deleteEmails('inbox', selectedEmails, pageNumber, emailsPerPage, totalEmails - selectedEmails.length, totalUnread - readEmails.length);
		},
		readEmails(selectedEmails) {
			const newEmails = emails.filter(({ id }) => selectedEmails.includes(id + '')).map(emailDetails => ({ ...emailDetails, isRead: true }));
			readEmails('inbox', newEmails, pageNumber, emailsPerPage, totalEmails, totalUnread - newEmails.length);
		}
	};
};

export default connect(
	mapStateToProps,
	{ fetchInboxDetails, fetchEmails, deleteEmails, readEmails },
	mergeProps
)(Inbox);
