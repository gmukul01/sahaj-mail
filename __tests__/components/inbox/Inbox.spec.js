import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Inbox from 'components/inbox/Inbox';

describe('Inbox Component', () => {
	const mockFetchEmail = jest.fn(),
		mockDeleteEmail = jest.fn(),
		mockReadEmail = jest.fn(),
		initialProps = {
			totalUnread: 2,
			emails: [],
			fetchEmails: mockFetchEmail,
			deleteEmails: mockDeleteEmail,
			readEmails: mockReadEmail
		};

	afterEach(() => {
		mockFetchEmail.mockClear();
		mockDeleteEmail.mockClear();
		mockReadEmail.mockClear();
	});

	it('should match snapshot', () => {
		const inbox = renderer.create(<Inbox {...initialProps} />).toJSON();
		expect(inbox).toMatchSnapshot();
	});

	it('should call fetchEmails when refresh is clicked', () => {
		const inbox = shallow(<Inbox {...initialProps} />);
		inbox.find('ActionBar').prop('onRefresh')();
		expect(mockFetchEmail.mock.calls.length).toBe(1);
	});

	it('should call deleteEmails when delete is clicked', () => {
		const inbox = shallow(<Inbox {...initialProps} />);
		inbox.find('ActionBar').prop('onDelete')();
		expect(mockDeleteEmail.mock.calls.length).toBe(1);
	});

	it('should call onRead when read is clicked', () => {
		const inbox = shallow(<Inbox {...initialProps} />);
		inbox.find('ActionBar').prop('onRead')();
		expect(mockReadEmail.mock.calls.length).toBe(1);
	});
});
