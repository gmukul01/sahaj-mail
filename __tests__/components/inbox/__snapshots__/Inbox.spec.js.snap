// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inbox Component should match snapshot 1`] = `
<section
  className="content inbox"
>
  <div
    className="inbox-topbar"
  >
    <div
      className="title"
    >
      Inbox ()
    </div>
  </div>
  <div
    className="inbox-actionbar"
  >
    <div
      className="actions"
    >
      <button
        className="button-outlined actions-refresh"
        onClick={[Function]}
      >
        <i
          className="fas fa-sync-alt"
        />
        Refresh
      </button>
      <button
        className="button-outlined"
        onClick={[Function]}
      >
        <i
          className="far fa-eye"
        />
      </button>
      <button
        className="button-outlined"
      >
        <i
          className="fas fa-exclamation"
        />
      </button>
      <button
        className="button-outlined"
        onClick={[Function]}
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
    </div>
    <div>
      <button
        className="button-outlined"
      >
        <i
          className="fas fa-arrow-left"
        />
      </button>
      <button
        className="button-outlined"
      >
        <i
          className="fas fa-arrow-right"
        />
      </button>
    </div>
  </div>
  <div
    className="email-list"
  />
</section>
`;
